In [5]:
import control
from control.matlab import *
import numpy as np
import scipy as sc

# Pregunta 3.1
$$
P_{2} = s^{6} + 8s^{5} + 23s^{4} + 36s^{3} + 38s^{2}+28s+16
$$

In [2]:
p1 = np.array([1,8,23,36,38,28,16])
np.roots(p1)

array([-4.00000000e+00+0.j, -2.00000000e+00+0.j, -1.00000000e+00+1.j,
       -1.00000000e+00-1.j,  6.24500451e-17+1.j,  6.24500451e-17-1.j])

# Pregunta 3.2

In [7]:
p2 = [1,-4,5,2-4j,2+4j,5]
ans = np.array([1])
for i in range(0,6):
    ans = sc.signal.convolve(ans,np.array([1,-p2[i]])) 
print(ans)

[ 1.0e+00+0.j -1.1e+01+0.j  3.9e+01+0.j  1.1e+01+0.j -7.4e+02+0.j
  2.7e+03+0.j -2.0e+03+0.j]


$$
s^{6} - 11s^{5} + 39s^{4} + 11s^{3} - 740s^{2} + 2700s - 2000
$$

# Pregunta 3.3

In [9]:
p1 = np.array([1,12,47,60])
p2 = np.array([4,28,83,135,126,62,12])
transfer_function = tf(p1,p2)
print(transfer_function)


                s^3 + 12 s^2 + 47 s + 60
-------------------------------------------------------
4 s^6 + 28 s^5 + 83 s^4 + 135 s^3 + 126 s^2 + 62 s + 12



In [10]:
pole(transfer_function)

array([-2. +0.j, -1. +1.j, -1. -1.j, -1.5+0.j, -1. +0.j, -0.5+0.j])

In [11]:
zero(transfer_function)

array([-5., -4., -3.])

# Pregunta 3.4


In [13]:
sc.signal.residue(p1,p2)

(array([ 10.5-1.24344979e-15j, -24. -2.13162821e-14j,
          2. +5.00000000e-01j,   2. -5.00000000e-01j,
         10.5+1.11910481e-14j,  -1. -4.44089210e-16j]),
 array([-0.5+0.j, -1. +0.j, -1. +1.j, -1. -1.j, -1.5+0.j, -2. +0.j]),
 array([], dtype=float64))

$$
\frac{P_{1}}{P_{2}} = \frac{10.5}{s+0.5} +\frac{-24}{s+1}+ \frac{2+0.5j}{s+1-j}+ \frac{2-0.5j}{s+1+j}+\frac{10.5}{s+1.5}+\frac{-1}{s+2}
$$

# Pregunta 3.5

In [16]:
n_values = [[1,1],[1,0,2],[1,4],[1,10]]
d_values = [[1,0],[1,2],[1,2,5],[2,1,4]]
p1 = np.array([1])
p2 = np.array([1])
for i in range(len(n_values)):
    polynom = np.array(n_values[i])
    p1 = sc.signal.convolve(p1,polynom)

for i in range(len(d_values)):
    polynom = np.array(d_values[i])
    p2 = sc.signal.convolve(p2,polynom)

transfer_function = tf(p1,p2)
print(transfer_function)


  s^5 + 15 s^4 + 56 s^3 + 70 s^2 + 108 s + 80
-----------------------------------------------
2 s^6 + 9 s^5 + 26 s^4 + 45 s^3 + 46 s^2 + 40 s



In [17]:
tf2ss(transfer_function)

A = [[-4.50000000e+00 -1.30000000e+00 -2.25000000e+00  2.30000000e+00
  -2.00000000e+00 -1.77635684e-15]
 [ 1.00000000e+01 -4.72227582e-16 -8.52827909e-17  1.29655032e-15
  -9.87270021e-16 -3.16267966e-17]
 [ 0.00000000e+00  1.00000000e+00  3.54919833e-17 -3.24470333e-17
   2.62203826e-17  1.93007607e-16]
 [ 0.00000000e+00  0.00000000e+00 -1.00000000e+00  8.18091527e-16
  -6.16837548e-16 -4.13856590e-16]
 [ 0.00000000e+00  0.00000000e+00  0.00000000e+00 -1.00000000e+00
  -8.32827089e-16  1.45959706e-16]
 [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00
   1.00000000e+00  2.22044605e-16]]

B = [[1.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]]

C = [[ 0.5   0.75  2.8  -3.5   5.4   4.  ]]

D = [[0.]]